### ISSUE TITLE
error with tor2web "Unhandled error in Deferrend"

### ISSUE THREAD

**Issue by catinova** [globaleaks/Tor2web#322]
I've this error after /etc/init.d/tor2web start

Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.spread.pb.PBConnectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.spread.pb.PBConnectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.spread.pb.PBConnectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.spread.pb.PBConnectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.spread.pb.PBConnectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Unhandled error in Deferred:
Unhandled Error
Traceback (most recent call last):
Failure: twisted.spread.pb.PBConnectionLost: [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionDone'>: Connection was closed cleanly.
]
Unhandled error in Deferred:
Unhandled Error

**Comment by catinova** [comment]
There isn't service on port 80. this is configuration file:

[main]

## TOR2WEB CONFIGURATION FOR GLOBALEAKS
# This is a minimal configuration template for a Tor2web to be used with
# globaleaks installed on the same server. !WARNING!: That setup is only
# valid for mid/low-risks deployments, such as anticorruption compliance in
# western countries

# Hostname of the public website
basehost = xxxxxxxxxxxxxxxx.it

# Local IP address Tor2web must be listening
listen_ipv4 = local ip of my DMZ

# Tor Hidden Service
# To find the hostname issue the following command
# cat /var/globaleaks/torhs/hostname
mode = TRANSLATION
onion =xxxxxxx.onion

# This allows Tor2web to work as a proxy for a local GlobaLeaks platform
# exposed on http://127.0.0.1:8082
dummyproxy = http://127.0.0.1:8082

# CERTIFICATE CONFIGURATION
# For the commands needed for generating the files below specified refer to the
# wiki page at: https://github.com/globaleaks/Tor2web/wiki/Installation-Guide
ssl_key = /home/tor2web/certs/tor2web-key.pem
#
# LOAD THE .CRT FILE THAT HAS BEEN GIVEN YOU BY YOUR CA
ssl_cert = /home/tor2web/certs/tor2web-cert.pem
#
# LOAD THE INTERMEDIATE CERTIFICATE CHAIN OF YOUR CA HERE
# PLEASE BE CAREFUL TO PUT THE CERT CHAIN IN THE RIGHT ORDER!
ssl_intermediate = /home/tor2web/certs/tor2web-intermediate.pem

# If globaleaks is used in embedded mode uncomment the settings below
# and read https://github.com/globaleaks/GlobaLeaks/wiki/Integration-Guide
# disable_tor_redirection = True
# blockhotlinking = False

# If you want to serve torbrowser files from this server
# uncomment this and read https://github.com/globaleaks/Tor2web/wiki/GetTor
disable_gettor = True

# If set to True will disable the tor2web disclaimer
disable_disclaimer = True

**Comment by evilaliv3** [comment]
thank you @catinova for reporting this.

whoud you please check with chown if the certificates are owned by the user/group tor2web and if this user has read access?

for the kind of configuration that you are doing you may be interested in taking track of the updates that we are doing in GlobaLeaks in order to integrate HTTPS and all the tor2web functionalities directly into it.

**Comment by catinova** [comment]
I've verified and permissions seem ok.
but the erroe it's always the same.

Any ideas ?

**Comment by chepurko** [comment]
I think I found a solution for this error, it seems like `tor2web` is looking for a DH parameters file but the GlobaLeaks wiki doesn't include that step. All I had to do was something like this:
```
cd /home/tor2web/certs/
openssl dhparam -out tor2web-dh.pem 2048
```
And also `chown` that file to the tor2web user and group.

**Comment by evilaliv3** [comment]
you are right @chepurko. \cc @fpietrosanti 

i will in a bit ship out a patch that will generate the DH parameters automagically also for tor2web so that the user will not have to issue manually any command in relation to this.

thanks for spotting the guide issue!

**Comment by chepurko** [comment]
@evilaliv3 Cool. Don't you think the defaults for DH and Let's Encrypt should be to 4,096 bits? All it would require from the `letsencrypt` step is the flag `--rsa-key-size 4096`. Their default is 2,048 bits.

**Comment by evilaliv3** [comment]
@chepurko this is a topic a lot discussed.

the security of a key (and in this case of the DH parameters) does not scale linearly with the number of bits so that it's generally preferramble to keep the key of 2048 to not impact on the performance of the system because anyhow the increased securrity of a key of 4096 bits is not that much.

You will find it discussed in a lot of autoritative software like gnupg  (https://www.gnupg.org/faq/gnupg-faq.html#no_default_of_rsa4096)  and others.

if you have any doubt feel free to ask.

**Comment by chepurko** [comment]
@evilaliv3 True, which is why I tried doing an ECDSA 384-bit key but had a problem with Let's Encrypt  accepting the CSR. This might have had something to do with Trusty's outdated OpenSSL 1.0.1f version because I didn't have a problem doing it on my Ubuntu 16.04 servers, but that's also another topic.
